# Day 5: Alchemical Reduction
[link](https://adventofcode.com/2018/day/5)

## Part 1: Reduce polymer

In [1]:
def reduce_poly(polymer, debug=False):
  p,i = list(polymer), 0
  while i < len(p)-1:
    if p[i] == p[i+1].swapcase():
      if debug:
        print(f'{to_s(p[:i])}({to_s(p[i:i+2])}){to_s(p[i+2:])}')
      del p[i]
      del p[i]
      i -= 1
    else:
      i += 1
  return to_s(p)

def to_s(array):
  return ''.join(array)

In [2]:
def assert_equal(actual, expected):
  assert actual == expected, f'Expected: "{expected}", got: "{actual}"'

assert_equal(reduce_poly('aA', True), '')
assert_equal(reduce_poly('abBA', True), '')
assert_equal(reduce_poly('abAB', True), 'abAB')

test_poly = 'dabAcCaCBAcCcaDA'
assert_equal(reduce_poly(test_poly, True), 'dabCBAcaDA')

(aA)
a(bB)A
(aA)
dabA(cC)aCBAcCcaDA
dab(Aa)CBAcCcaDA
dabCBA(cC)caDA


In [3]:
with open('05 input.txt', 'r') as file:
  puzzle_input = file.readline().strip()

display(len(puzzle_input))

part_1_reduced = reduce_poly(puzzle_input)
len(part_1_reduced)

50000

9390

**Part 1 correct answer:** `9390`

## Part 2: Remove a unit type to maximize polymer reduction

In [4]:
import re

def remove_unit(polymer, unit):
  assert len(unit) == 1
  return re.sub(r'[' + unit + unit.swapcase() + ']', '', polymer)

assert_equal(remove_unit(test_poly, 'a'), 'dbcCCBcCcD')
assert_equal(remove_unit(test_poly, 'A'), 'dbcCCBcCcD')
assert_equal(remove_unit(test_poly, 'b'), 'daAcCaCAcCcaDA')

In [5]:
def remove_unit_reduce(polymer, debug=False):
  units = sorted(set(polymer.lower()))
  if(debug):
    print(f'Polymer:\t{polymer}')
  print(f'Units ({len()}):\t\t{"".join(units)}')

  lengths = []
  for unit in units:
    poly_unit_removed = remove_unit(polymer, unit)
    print(f'\nRemoved unit {unit}/{unit.swapcase()}')
    if(debug):
      print('Polymer: {poly_unit_removed}')
    
    poly_reduced = reduce_poly(poly_unit_removed, debug)
    if(debug):
      print(f'Reduced polymer: {poly_reduced}')
    print(f'Reduced length: {len(poly_reduced)}')
    
    lengths.append(len(poly_reduced))

  return min(lengths)

assert_equal(remove_unit_reduce(test_poly, True), 4)

Polymer:	dabAcCaCBAcCcaDA
Units:		abcd

Removed unit a/A
Polymer: {poly_unit_removed}
db(cC)CBcCcD
dbCB(cC)cD
Reduced polymer: dbCBcD
Reduced length: 6

Removed unit b/B
Polymer: {poly_unit_removed}
d(aA)cCaCAcCcaDA
d(cC)aCAcCcaDA
daCA(cC)caDA
Reduced polymer: daCAcaDA
Reduced length: 8

Removed unit c/C
Polymer: {poly_unit_removed}
dab(Aa)BAaDA
da(bB)AaDA
d(aA)aDA
Reduced polymer: daDA
Reduced length: 4

Removed unit d/D
Polymer: {poly_unit_removed}
abA(cC)aCBAcCcaA
ab(Aa)CBAcCcaA
abCBA(cC)caA
abCBAc(aA)
Reduced polymer: abCBAc
Reduced length: 6


In [6]:
remove_unit_reduce(puzzle_input)

Units:		abcdefghijklmnopqrstuvwxyz

Removed unit a/A
Reduced length: 9036

Removed unit b/B
Reduced length: 9026

Removed unit c/C
Reduced length: 9022

Removed unit d/D
Reduced length: 8998

Removed unit e/E
Reduced length: 9036

Removed unit f/F
Reduced length: 8988

Removed unit g/G
Reduced length: 9018

Removed unit h/H
Reduced length: 9014

Removed unit i/I
Reduced length: 8998

Removed unit j/J
Reduced length: 9000

Removed unit k/K
Reduced length: 8980

Removed unit l/L
Reduced length: 8964

Removed unit m/M
Reduced length: 9044

Removed unit n/N
Reduced length: 9052

Removed unit o/O
Reduced length: 9064

Removed unit p/P
Reduced length: 9010

Removed unit q/Q
Reduced length: 9034

Removed unit r/R
Reduced length: 9026

Removed unit s/S
Reduced length: 9016

Removed unit t/T
Reduced length: 9002

Removed unit u/U
Reduced length: 5898

Removed unit v/V
Reduced length: 8992

Removed unit w/W
Reduced length: 8950

Removed unit x/X
Reduced length: 9028

Removed unit y/Y
Reduced len

5898

**Part 2 correct answer:** `5898`